src/thunderbots/software/multithreading/thread_safe_buffer.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <boost/circular_buffer.hpp>
+#include <condition_variable>
+#include <cstddef>
+#include <deque>
+#include <mutex>
+#include <optional>
+
+#include "util/time/duration.h"
+
+/**
+ * This class represents a buffer of objects
+ *
+ * It's public API is fully thread-safe (meaning that it can be accessed by
+ * multiple threads at the same time without issue)
+ *
+ * @tparam T The type of whatever is being buffered
+ */
+template <typename T>
+class ThreadSafeBuffer
+{
+   public:
+    // Force the user to specify a size
+    explicit ThreadSafeBuffer() = delete;
+
+    explicit ThreadSafeBuffer(std::size_t buffer_size);
+
+    // Copying this class is not permitted
+    ThreadSafeBuffer(const ThreadSafeBuffer&) = delete;
+
+    /**
+     * Removes the least recently value added to the buffer and returns it
```

Comment:
Changed all instances of `least recently` to `oldest`

Reply:
Hmm, so I was trying to avoid words like "oldest", because we have some types with associated timestamps, and I wanted to avoid confusion. The intent here is that if I add "A,B,C" to this buffer (in that order), then this would return "A". I see how this naming could be confusing though. Any more suggestions?

Reply:
I don't fully understand your concern with types that have associated timestamps?

Because this is a buffer maybe we could use "front" and "back" terminology? `popFront` or `popBack` maybe would be more familiar to people?

Reply:
Well if I a have a `ThreadSafeBuffer<World>` with a `popMostRecent`, then that could be the `World` with the largest timestamp, or the world that I most recently added to the buffer.

I don't think front and back is a great idea, what is "front" and "back"? Intuitively I'd have not idea what's "newer".

Reply:
I like the idea of using timestamps, front/back isn't as intuitive. I like popMostRecent()

Reply:
I would also be fine if you changed the comment to be clearer

Reply:
But the buffer isn't making guarantees about returning something with the "most recent timestamp". If for whatever reason the "most recent" item (aka the one with the largest timestamp) isn't at the front of the buffer, it is not going to get returned (whatever is at the front will be returned).

Because the buffer class doesn't enforce timestamp ordering, I think the terminology should be more "positional" than "time based", if that makes sense. `getMostRecent` is somewhat ok, but like mentioned above can be taken to mean "time". Maybe `getMostRecentlyAdded`.

Reply:
Actually I guess that means I prefer the current naming. Like Evan said, maybe make the comments slightly more clear, but overall I like what we have now.

I would use `popLeastRecentlyAddedValue` because "pop" is the most standard way to indicate something is also removed, as opposed to "pull". IMO `pop` or `get` would be preferred here.

Reply:
Added a quick example to the javadoc comment: `ex. if A,B,C were added to the buffer (in that order), this would return A` and renamed to `popLeastRecentlyAddedValue`

Reply:
Would we not be trying to grab the most recent data 95% of the time?

Reply:
If it's timestamp ordered then this wouldn't be a generic class. I also think that having a buffer that's agnostic of what we're putting in it could be useful. Consider the case where we pass data that's not timestamped.